### PR TITLE
Eth1 Integration

### LINKED ISSUE #463: Implement Eth1 RPC linking
## Description

Beacon nodes must connect to Eth1 nodes and gather (at a minimum):

- Logs emitted by the deposit contract.
- The values of the `get_hash_tree_root()` and `get_deposit_count()` calls on the deposit contract.
- The hash of some block at some height in the Eth1 chain.

## Steps to resolve

@pawanjay176 is working on this and we have discussed a lot offline.

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

- Resolves #463
- Supersedes #491

## Proposed Changes

- Add functions to connect to an `eth1` node to get relevant the `Eth1Data` using the `web3` crate.
- Add `get_eth1_vote` function to determine new state for Beacon chain `state.eth1_data` using the above data.
- Addresses a bunch of clippy lints across the entire project.
  - There are also quite a few modifications from `rustfmt`
- Allows the genesis `Fork` to be defined in `ChainSpec`.
- Ensures the `Fork` for  verifying `Deposit` is always a constant `[0, 0, 0, 0]`.
  - Removes some unnecessary parameters from the deposit `SignatureSet`.
- Unifies the `beacon_node` and `validator_client` into a single `ligyhthouse` binary.
  - Introduces the concept of an `Environment` and a `Context`, which provide access to tokio and logging.
- Adds basic framework for testing a `BeaconNode` in-process
  - Allows for services to be started on port `0` (i.e., let the OS choose a free port) and read which port was actually used.

### CODE REVIEW

**Comment by paulhauner** [comment]
Note: before merging ensure that the image for gitlab ci is set back to `latest` (not `eth1`)

**Comment by paulhauner** [comment]
Blocked on https://github.com/ethereum/eth2.0-specs/issues/1464

File: beacon_node/genesis/src/common.rs [hunk 1]
```diff
@@ -0,0 +1,50 @@
+use merkle_proof::MerkleTree;
+use rayon::prelude::*;
+use tree_hash::TreeHash;
+use types::{ChainSpec, Deposit, DepositData, Hash256};
+
+/// Accepts the genesis block validator `DepositData` list and produces a list of `Deposit`, with
+/// proofs.
+pub fn genesis_deposits(deposit_data: Vec<DepositData>, spec: &ChainSpec) -> Vec<Deposit> {
```
Comment (paulhauner) [inline beacon_node/genesis/src/common.rs]: TODO: Merge in master and update with Scott's more efficient method.

File: beacon_node/genesis/src/eth1_genesis_service.rs [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+                .try_for_each(|deposit| {
+                    // No need to verify proofs in order to test if some block will trigger genesis.
+                    const PROOF_VERIFICATION: bool = false;
+
+                    process_deposit(
+                        &mut local_state,
+                        &deposit,
+                        spec,
+                        PROOF_VERIFICATION,
+                        // TODO: disable signature verification
```
Comment (paulhauner) [inline beacon_node/genesis/src/eth1_genesis_service.rs]: This could be made much faster by pre-verifying block signatures, filtering out those with bad signatures, then disabling signature verification in this function.

File: beacon_node/genesis/src/eth1_genesis_service.rs [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+            .iter()
+            // It's only worth scanning blocks that have timestamps _after_ genesis time. It's
+            // impossible for any other block to trigger genesis.
+            .filter(|block| block.timestamp >= spec.min_genesis_time)
+            // The block cache might be more recently updated than deposit cache. Restrict any
+            // block numbers that are not known by all caches.
+            .filter(|block| {
+                self.highest_known_block()
+                    .map(|n| block.number <= n)
+                    .unwrap_or_else(|| true)
```
Comment (paulhauner) [inline beacon_node/genesis/src/eth1_genesis_service.rs]: Should this be `false` instead?

File: beacon_node/genesis/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+use ssz::Decode;
+use std::fs::File;
+use std::io::prelude::*;
+use std::path::PathBuf;
+use std::time::Duration;
+use types::{BeaconState, ChainSpec, EthSpec};
+
+const ETH1_GENESIS_UPDATE_INTERVAL_MILLIS: u64 = 500;
+
+pub enum Eth2Genesis {
```
Comment (paulhauner) [inline beacon_node/genesis/src/lib.rs]: Perhaps the `ClientStartMethod` (or w/e) should just embed this?

File: beacon_node/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    /// Identical to `start_from_client_config`, however the `client_config` is generated from the
+    /// given `matches` and potentially configuration files on the local filesystem or other
+    /// configurations hosted remotely.
+    pub fn new_from_cli<'a, 'b>(
+        mut context: RuntimeContext<E>,
+        matches: &ArgMatches<'b>,
+    ) -> impl Future<Item = Self, Error = String> + 'a {
+        let log = context.log.clone();
+
+        // FIXME: the eth2 config in the env is being completely ignored.
```
Comment (paulhauner) [inline beacon_node/src/lib.rs]: TODO: this

File: eth2/utils/compare_fields_derive/src/lib.rs [hunk 6]
```diff
@@ -6,12 +6,10 @@ use quote::quote;
 use syn::{parse_macro_input, DeriveInput};
 
 fn is_slice(field: &syn::Field) -> bool {
-    for attr in &field.attrs {
-        if attr.tts.to_string() == "( as_slice )" {
-            return true;
-        }
-    }
-    false
+    field.attrs.iter().any(|attr| {
```
Comment (paulhauner) [inline eth2/utils/compare_fields_derive/src/lib.rs]: I'm not sure why these changes are here..

File: eth2/utils/eth2_config/src/lib.rs [hunk 7]
```diff
@@ -53,26 +52,14 @@ impl Eth2Config {
     /// invalid.
     pub fn apply_cli_args(&mut self, args: &ArgMatches) -> Result<(), &'static str> {
         if args.is_present("recent-genesis") {
-            self.spec.min_genesis_time = recent_genesis_time()
+            // FIXME: re-enable this.
```
Comment (paulhauner) [inline eth2/utils/eth2_config/src/lib.rs]: TODO: this

File: eth2/utils/ssz_derive/src/lib.rs [hunk 8]
```diff
@@ -53,12 +53,9 @@ fn get_serializable_field_types<'a>(struct_data: &'a syn::DataStruct) -> Vec<&'a
 ///
 /// The field attribute is: `#[ssz(skip_serializing)]`
 fn should_skip_serializing(field: &syn::Field) -> bool {
-    for attr in &field.attrs {
-        if attr.tts.to_string() == "( skip_serializing )" {
-            return true;
-        }
-    }
-    false
+    field.attrs.iter().any(|attr| {
```
Comment (paulhauner) [inline eth2/utils/ssz_derive/src/lib.rs]: Not sure why this is here.

File: eth2/utils/test_random_derive/src/lib.rs [hunk 9]
```diff
@@ -9,12 +9,9 @@ use syn::{parse_macro_input, DeriveInput};
 ///
 /// The field attribute is: `#[test_random(default)]`
 fn should_use_default(field: &syn::Field) -> bool {
-    for attr in &field.attrs {
-        if attr.tts.to_string() == "( default )" {
-            return true;
-        }
-    }
-    false
+    field.attrs.iter().any(|attr| {
```
Comment (paulhauner) [inline eth2/utils/test_random_derive/src/lib.rs]: Not sure why this is here.

File: tests/eth1_test_rig/contract/v0.8.3_validator_registration.bytecode [hunk 10]
```diff
@@ -0,0 +1 @@
...0208501600060046015f150508051820191505080610b2052610b209050602060c0825160208401600060025af1610f7057600080fd5b60c0519050602082610ca00101526020810190506000610360600880602084610c2001018260208501600060046012f15050805182019150506000601860208206610ba00160208284011115610fc557600080fd5b602080610bc082610720600060046015f1505081815280905090509050601880602084610c2001018260208501600060046014f150508051820191505061086051602082610c2001015260208101905080610c2052610c209050602060c0825160208401600060025af161103857600080fd5b60c0519050602082610ca001015260208101905080610ca052610ca09050602060c0825160208401600060025af161106f57600080fd5b60c0519050610b0052600180546001825401101561108c57600080fd5b6001815401815550600154610d2052610d4060006020818352015b60016001610d20511614156110dc57610b0051610d4051602081106110cb57600080fd5b600060c052602060c0200155611170565b6000610d4051602081106110ef57600080fd5b600060c052602060c0200154602082610d60010152602081019050610b0051602082610d6001015260208101905080610d6052610d609050602060c0825160208401600060025af161114057600080fd5b60c0519050610b0052610d20600261115757600080fd5b60028151048152505b81516001018083528114156110a7575b5050005b60006000fd5b61017f6112f90361017f60003961017f6112f9036000f3"
```
Comment (paulhauner) [inline tests/eth1_test_rig/contract/v0.8.3_validator_registration.bytecode]: Should this be in the repo?

File: tests/eth1_test_rig/contract/v0.8.3_validator_registration.json [hunk 11]
```diff
@@ -0,0 +1 @@
+[{"anonymous":false,"inputs":[{"indexed":false,"name":"pubkey","type":"bytes"},{"indexed":false,"name":"withdrawal_credentials","type":"bytes"},{"indexed":false,"name":"amount","type":"bytes"},{"indexed":false,"name":"signature","type":"bytes"},{"indexed":false,"name":"index","type":"bytes"}],"name":"DepositEvent","type":"event"},{"constant":false,"inputs":[],"outputs":[],"payable":false,"type":"constructor"},{"constant":true,"gas":91674,"inputs":[],"name":"get_hash_tree_root","outputs":[{"name":"out","type":"bytes32"}],"payable":false,"type":"function"},{"constant":true,"gas":10433,"inputs":[],"name":"get_deposit_count","outputs":[{"name":"out","type":"bytes"}],"payable":false,"type":"function"},{"constant":false,"gas":1334417,"inputs":[{"name":"pubkey","type":"bytes"},{"name":"withdrawal_credentials","type":"bytes"},{"name":"signature","type":"bytes"}],"name":"deposit","outputs":[],"payable":true,"type":"function"}]
```
Comment (paulhauner) [inline tests/eth1_test_rig/contract/v0.8.3_validator_registration.json]: Should this be in the repo?

**Comment by paulhauner** [comment]
This is finally ready for review!

File: beacon_node/beacon_chain/src/eth1_chain.rs [hunk 12]
```diff
@@ -89,22 +147,977 @@ impl<T: EthSpec> Eth1ChainBackend<T> for InteropEth1ChainBackend<T> {
...
+            log: service.log.clone(),
+            core: service,
+            store,
             _phantom: PhantomData,
         }
     }
 }
 
+impl<T: EthSpec, S: Store> Eth1ChainBackend<T> for CachingEth1Backend<T, S> {
+    fn eth1_data(&self, state: &BeaconState<T>, spec: &ChainSpec) -> Result<Eth1Data, Error> {
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/eth1_chain.rs]: I've done my best to grok this, and I'm fairly sure it matches the spec: https://github.com/ethereum/eth2.0-specs/blob/v0.9.1/specs/validator/0_beacon-chain-validator.md#eth1-data

File: beacon_node/beacon_chain/src/eth1_chain.rs [hunk 13]
```diff
@@ -89,22 +147,977 @@ impl<T: EthSpec> Eth1ChainBackend<T> for InteropEth1ChainBackend<T> {
...
+        });
+
+    valid_votes
+}
+
+/// Selects the winning vote from `valid_votes`.
+fn find_winning_vote(valid_votes: Eth1DataVoteCount) -> Option<Eth1Data> {
+    valid_votes
+        .iter()
+        .max_by_key(|((_eth1_data, block_number), vote_count)| (*vote_count, block_number))
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/eth1_chain.rs]: Nice use of the `block_number` :sunglasses:

File: tests/eth1_test_rig/src/ganache.rs [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+            .stdout(Stdio::piped())
+            .arg("--defaultBalanceEther")
+            .arg("1000000000")
+            .arg("--gasLimit")
+            .arg("1000000000")
+            .arg("--accounts")
+            .arg("10")
+            .arg("--port")
+            .arg(format!("{}", port))
+            .arg("--seed")
```
Comment (michaelsproul) [inline tests/eth1_test_rig/src/ganache.rs]: Same as for the script, should probably be `--mnemonic`

File: beacon_node/eth1/src/deposit_log.rs [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+        let signature = bytes
+            .get(SIG_START..SIG_START + SIG_LEN)
+            .ok_or_else(|| "Insufficient bytes for signature".to_string())?;
+        let index = bytes
+            .get(INDEX_START..INDEX_START + INDEX_LEN)
+            .ok_or_else(|| "Insufficient bytes for index".to_string())?;
+
+        let deposit_data = DepositData {
+            pubkey: PublicKeyBytes::from_ssz_bytes(pubkey)
+                .map_err(|e| format!("Invalid index ssz: {:?}", e))?,
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply beacon_node/eth1/src/deposit_log.rs]: Thank you! I'm sure this would have been super annoying at some point.

File: beacon_node/beacon_chain/src/eth1_chain.rs (hunk 12 above)
Comment (paulhauner) [reply beacon_node/beacon_chain/src/eth1_chain.rs]: phew!

File: beacon_node/beacon_chain/src/eth1_chain.rs [hunk 16]
```diff
@@ -1,27 +1,90 @@
...
-    backend: T::Eth1Chain,
+pub struct Eth1Chain<T, E>
+where
+    T: Eth1ChainBackend<E>,
+    E: EthSpec,
+{
+    backend: T,
+    /// When `true`, the backend will be ignored and dummy data from the 2019 Canada interop method
+    /// will be used instead.
+    pub use_dummy_backend: bool,
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply beacon_node/beacon_chain/src/eth1_chain.rs]: I totally agree with you conceptually and originally I had it the way you suggested. _But_, it turns out that including it at the type level requires significantly more code on the run-time side and this hack seemed like a fair trade off to me.

Most of my reasoning is that I don't really see a reason for `DummyEth1Backend` to exist; voting on deterministic not-real-data in order to keep the deposit count stable is equivalent to just voting completely random junk. That being said, I'm not quite sure I'm ready to raise this up the "client interop" flagpole yet, so I thought we'd leave it in here for now.

We can chat about this in person today, but my feeling is to leave this as-is for now instead of (a) spending the time on a refactor and (b) introducing more code that isn't run during tests but is used at run-time.
